Cleanup token algebra

carlostome opened the pull request:
# Description

This PR aims to organize and instantiate TokenAlgebra (possibly solving #711).

In particular,

- Moves Coin instance of TA to its own module.

~- Removes the less-or-equals than relation from TokenAlgebra.~

~This relation is used by the UTxO rule to check that every output of a transaction has at least `(overhead + utxoEntrySize txout) * coinsPerUTxOByte pp` coin. The way this check is currently performed is by comparing the coin at the level of Value, via injection. This relies somewhat implicitly on two things: `inject` does the "right thing" for assets other than Ada; and the less-or-equals relation for TA agrees with that of Coin. These requirements are not part of the mechanization.~

~Another issue is to define what kind of relation should the less-or-equals be. Since Value (as per the shelley-ma formal spec.) is meant to be an instance, the relation can not be a total order but at most a pre/partialorder.~

~To side-step these shortcomings, this PR proposes to remove the relation altogether from TA and adapt the check in the UTxO rule.~

- Removes `AssetName` and `specialAsset` from TokenAlgebra

These fields are not necessary since by construction Coin is a retract of TokenAlgebra, via `inject` and `coin`, and hence we already know how to "extract" Coin from a TA.
 
- Rename `property`

- Add instance for fixed number of multi-assets (in `ValueVector`)

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] Any semantic changes to the specifications are documented in `CHANGELOG.md`
- [x] Code is formatted according to [CONTRIBUTING.md](https://github.com/input-output-hk/formal-ledger-specifications/blob/master/CONTRIBUTING.md)
- [x] Self-reviewed the diff

WhatisRT commented on src/Ledger/Conway/Conformance/Utxo.agda:
```diff
@@ -95,7 +95,7 @@ data _⊢_⇀⦇_,UTXO⦈_ : UTxOEnv → UTxOState → Tx → UTxOState → Type
 
     ∙ ∀[ (_ , txout) ∈ txoutsʰ .proj₁ ]
-        inject ((overhead + L.utxoEntrySize txout) * coinsPerUTxOByte pp) ≤ᵗ getValueʰ txout
+        ((overhead + L.utxoEntrySize txout) * coinsPerUTxOByte pp) ≤ coin (getValueʰ txout)
```
This isn't equivalent to the previous line, since that one also ensures that no negative amounts are put in an output.

We either need a `TokenAlgebra` instance with negative amounts (at least being supported) because `mint` must be able to contain them to burn tokens, or we have to do what the implementation does (I think) which is to have a bunch of different types. I much prefer having this inequality over another type.

WhatisRT commented:
> Another issue is to define what kind of relation should the less-or-equals be. Since Value (as per the shelley-ma formal spec.) is meant to be an instance, the relation can not be a total order but at most a pre/partialorder.

That's fine, we have a `HasPreorder` typeclass that can be used here.

carlostome commented:
I reverted the removal of ≤ᵗ from this PR, and changed `VectorValue` accordingly.

The discussion about removing ≤ᵗ from Token Algebra has been moved to its own issue (#761)

WhatisRT commented on src/Ledger/TokenAlgebra/ValueVector.agda:
```diff
@@ -0,0 +1,50 @@
+module Vec-commutativeMonoid = Alg.CommutativeMonoid Vec-commutativeMonoid
+
+module _ (Policies : Vec PolicyId n) where
+  Value-TokenAlgebra : TokenAlgebra
```
I think it would be good to also implement the FFI for this type in this PR, since that seems to be the main point of this particular algebra. I'd say it should map to a `Coin × List (PolicyId × String × ℕ)`, and the conversion to the Agda type should error if any of the `String`s are nonempty or there are any duplicate `PolicyId`s.

WhatisRT reviewed (approved):
LGTM, I've approved it in case you want to merge it now and have other plans for the FFI

WhatisRT commented:
Ah, I didn't see the auto merge, I guess it'll be future work then
